### ISSUE TITLE
research  - vast repository of technologies

### ISSUE THREAD

**Issue by fullstackslayer** [hydralite/hydralite#60]
A lot of our discovery features depend on the technologies a user has on their profile.
One example of this would be recommending projects to open source developers based on the technologies they work with.

We need some sort of a large collection of technologies, similar to how StackOverflow and GitHub have.

It would be ideal to store all these technologies in our database or so, to make relating data easier.

**Comment by DudeBro249** [comment]
Found a good resource - https://github.com/stackshareio/awesome-stacks

**Comment by ShubhamPatilsd** [comment]
Could you elaborate a bit more about what you mean?

**Comment by fullstackslayer** [comment]
> Could you elaborate a bit more about what you mean?

@ShubhamPatilsd when a user signs up, we have an onboarding process that lets them choose the technologies they work with. This enhances our project discovery feature so they only get recommend project that have a tech stack that matches their skillset. 

The issue at hand is, where do we get all these technologies for them to choose from.

**Comment by XtremeDevX** [comment]
Perhaps we hard code it for now? Don't think there would be api's for this kinda thing.

**Comment by fullstackslayer** [comment]
> Perhaps we hard code it for now? Don't think there would be api's for this kinda thing.

that wouldn't be a smart thing to do, why manually do it ourselves when people have already compiled vast lists of them. Worst case scenario we make a pyautogui bot to type in random letters on say stack overflow in the tags section and just wwbscrape the technologies and throw them into our db 😂

**Comment by XtremeDevX** [comment]
🤣 lemme search for API's and see if I can find one

**Comment by fullstackslayer** [comment]
> 🤣 lemme search for API's and see if I can find one

sounds good

**Comment by XtremeDevX** [comment]
Right, so the StackOverflow API might help - we'll get access to all StackOverflow tags

**Comment by fullstackslayer** [comment]
> Right, so the StackOverflow API might help - we'll get access to all StackOverflow tags

oh wow, I didn't know the SO API would give you the tags

**Comment by fullstackslayer** [comment]
You are right, this looks amazing:
https://api.stackexchange.com/2.2/tags?page=1&order=desc&sort=popular&site=stackoverflow

I think we can run an iteration script and basically increment the `page` query param until it runs out of pages. on each iteration, we can run a sub iteration and load the tags into the db. Im assuming this is legal since its public data, but we need to confirm once before adding it to our database. Maybe we can attribute stack exchange somehow.

**Comment by fullstackslayer** [comment]
My concern is storing all of this, I did a bit of trial and error and noticed there were roughly 2,000 pages with each page having about 30 tags. That would make 60000 technologies we would need to store 😬...

We could perhaps directly use their API to query all the technologies instead of having to put them into our db...

We can also sort them by popularity, so maybe we just iterate over the first x amount of pages.

**Comment by XtremeDevX** [comment]
I'm more concerned about the fact that technologies are constantly changing

**Comment by fullstackslayer** [comment]
> I'm more concerned about the fact that technologies are constantly changing. Storing them would mean we would need to update them on a weekly basis.

How bout we allow our community to add technologies they don't see? For example, if I was a Lua developer and Lua wasn't a technology listed, I can submit it as a technology to be added and a hydralite team can review and approve it accordingly...

**Comment by XtremeDevX** [comment]
Also, 100 pages should be enough, and besides that, 3000 tags ain't that much, databases can handle that with EASE

**Comment by XtremeDevX** [comment]
but we have to start with an existing list

**Comment by fullstackslayer** [comment]
yeah

**Comment by fullstackslayer** [comment]
@XtremeDevX would you mind building an automation script to roughly extract these technologies from the stack exchange api? I can build you a mutation that you can send each technology to so we can start loading stuff in the db. Make sure to run sleep(500) after retrieving each technology or smth.

**Comment by fullstackslayer** [comment]
you can put the script under a directory called "automation"

**Comment by fullstackslayer** [comment]
follow up -
What I'm thinking is to allow them to retrieve tags from the stack overflow API itself. when they send the tags they chose to our API, we first check to see if we have already stored them in Redis. if we haven't, we add them to accordingly. this will make a good caching system where we don't need to store everything at once.

Give me your thoughts @XtremeDevX @ShubhamPatilsd @samarmohan

**Comment by VarunPotti** [comment]
Try to switch Redis w/ PostgreSQL
As we won't be needing the data each n every second. Redis might not be needed for this task.
